src/CoreWCF.Http/src/CoreWCF/Channels/WebSocketTransportDuplexSessionChannel.cs

```diff
@@ -753,8 +754,8 @@ private async Task<Message> PrepareMessageAsync(WebSocketReceiveResult result, b
                     }
                     else
                     {
-                        ArraySegment<byte> bytes = new ArraySegment<byte>(buffer, 0, count);
-                        message = _encoder.ReadMessage(bytes, _bufferManager);
+                        ReadOnlySequence<byte> bytes = new(buffer, 0, count);
+                        message = await _encoder.ReadMessageAsync(bytes, _bufferManager);
```

Comment:
I don't know if you want to do this here, but there's an opportunity for a perf improvement here. If you look at the code around line 520, the current code can make multiple calls to WebSocket.ReceiveAsync and if the EndOfMessage flag isn't set in the result, it calls it again. ReceiveAsync is passed an ArraySegment so if there's space in the buffer, it just passes the same buffer with the offset starting where the previous ReadAsync completed. But if the buffer is full, it allocates a new one twice the size and copies the previous once across. But when dealing with a ReadOnlySequence, you don't need a single buffer. That code around line 520 could allocate a new buffer the same size as the first one and fill that up. PrepareMessageAsync could then be passed the ReadOnlySequence containing all the buffers without any copying having occurred.

Reply:
@mconnew would you accept a follow up PR dedicated to the perf imrpovement ?